include/pybind11/pytypes.h

```diff
@@ -375,31 +388,101 @@ PYBIND11_NAMESPACE_END(detail)
+                        msg += "PYTHON_EXCEPTION_CLASS_NAME_IS_NULLPTR";
+                    } else {
+                        msg += class_name;
+                    }
+                }
+                msg += ": ";
+                PyObject *val_str = PyObject_Str(m_value.ptr());
+                if (val_str == nullptr) {
+                    msg += "PYTHON_EXCEPTION_VALUE_IS_NULLPTR";
+                } else {
+                    Py_ssize_t utf8_str_size = 0;
+                    const char *utf8_str = PyUnicode_AsUTF8AndSize(val_str, &utf8_str_size);
+                    if (utf8_str == nullptr) {
+                        msg += "PYTHON_EXCEPTION_VALUE_AS_UTF8_FAILURE";
+                    } else {
+                        msg += '"' + std::string(utf8_str, static_cast<std::size_t>(utf8_str_size))
+                               + '"';
+                    }
+                }
+                // Intentionally using C calls to maximize reliability
+                // (and to avoid #include <iostream>).
+                fprintf(stderr, "%s [STDERR]\n", msg.c_str());
+                fflush(stderr);
+                fprintf(stdout, "%s [STDOUT]\n", msg.c_str());
+                fflush(stdout);
+                std::terminate();
+            }
+        }
+        return m_lazy_what.c_str();
+    }
+
+    /// Restores the currently-held Python error (which will clear the Python error indicator first
+    /// if already set).
+    /// NOTE: This function will not necessarily restore the original Python exception, but may
+    ///       restore the normalized exception if what() or discard_as_unraisable() were called
+    ///       prior to restore().
     void restore() {
-        PyErr_Restore(m_type.release().ptr(), m_value.release().ptr(), m_trace.release().ptr());
+        // As long as this type is copyable, there is no point in releasing m_type, m_value,
+        // m_trace, but simply holding on the the references makes it possible to produce
+        // what() even after restore().
+        PyErr_Restore(m_type.inc_ref().ptr(), m_value.inc_ref().ptr(), m_trace.inc_ref().ptr());
     }
 
     /// If it is impossible to raise the currently-held error, such as in a destructor, we can
     /// write it out using Python's unraisable hook (`sys.unraisablehook`). The error context
     /// should be some object whose `repr()` helps identify the location of the error. Python
-    /// already knows the type and value of the error, so there is no need to repeat that. After
-    /// this call, the current object no longer stores the error variables, and neither does
-    /// Python.
+    /// already knows the type and value of the error, so there is no need to repeat that.
+    /// NOTE: This function may have the side-effect of normalizing the held Python exception
+    ///       (if it is not normalized already).
     void discard_as_unraisable(object err_context) {
+#if PY_VERSION_HEX < 0x03080000
```

Comment:
This complication is why I suggested just normalizing it in the CTOR initially, at least none of the method would have side effects.

Reply:
+ The benchmarked overhead for normalizing the object is only ~100ns .